### PR TITLE
Support non-standard bools in CUDA mode

### LINKED ISSUE #54789: Conversion from bool to float sometimes produces tensor with values 255.0 for True instead of 1.0
## 🐛 Bug

When I have two seemingly identical bool tensors with values: `tensor([True, True, True, True])`, one of them produces a float tensor with values: `tensor([1.0, 1.0, 1.0, 1.0])`. The other one produces a float tensor with values: `tensor([255., 255., 255., 255.])`.

## To Reproduce

```python
import numpy as np
import torch
# these are the same data and they look the same when you load them, but they behave subtly differently
poisoned = np.frombuffer(b'\xff\xff\xff\xff', dtype=np.bool) # hex 255
clean = np.frombuffer(b'\x01\x01\x01\x01', dtype=np.bool)

torch_poisoned = torch.from_numpy(poisoned)
torch_clean = torch.from_numpy(clean)


print('poisoned == clean -->', (poisoned == clean).all())  # >>> True
print('torch_poisoned == torch_clean --> ', (torch_poisoned == torch_clean).all())  # >>> tensor(False)

print('torch_poisoned.dtype == torch_clean.dtype --> ', (torch_poisoned.dtype == torch_clean.dtype))  # >>> True
print('torch_poisoned.dtype, torch_clean.dtype --> ', torch_poisoned.dtype, torch_clean.dtype)  # >>> torch.bool, torch.bool

print('torch_clean: ',torch_clean)  # >>> tensor([True, True, True, True])
print('torch_clean.float(): ', torch_clean.float())  # >>> tensor([1., 1., 1., 1.])

print('torch_poisoned: ', torch_poisoned)  # >>> tensor([True, True, True, True])
print('torch_poisoned.float(): ', torch_poisoned.float())  # >>> tensor([255., 255., 255., 255.])
```

## Expected behavior

Any thing that torch calls a True bool should be converted to a float value of 1.0, not whatever bytes happened to be stored inside the bool.

This could just be an abuse of bool, but I came across this bug when using converting a boolean image using PIL. PyTorch should probably prevent users from being able to shoot their foot like this.

## Environment

Collecting environment information...
PyTorch version: 1.7.1+cu110
Is debug build: False
CUDA used to build PyTorch: 11.0
ROCM used to build PyTorch: N/A

OS: Ubuntu 18.04.5 LTS (x86_64)
GCC version: (Ubuntu 7.5.0-3ubuntu1~18.04) 7.5.0
Clang version: Could not collect
CMake version: version 3.15.3

Python version: 3.6 (64-bit runtime)
Is CUDA available: True
CUDA runtime version: 11.2.142
GPU models and configuration: GPU 0: GeForce GTX 1080 Ti
Nvidia driver version: 455.45.01
cuDNN version: Probably one of the following:
/usr/lib/x86_64-linux-gnu/libcudnn.so.7.6.5
/usr/lib/x86_64-linux-gnu/libcudnn.so.8.1.1
/usr/lib/x86_64-linux-gnu/libcudnn_adv_infer.so.8.1.1
/usr/lib/x86_64-linux-gnu/libcudnn_adv_train.so.8.1.1
/usr/lib/x86_64-linux-gnu/libcudnn_cnn_infer.so.8.1.1
/usr/lib/x86_64-linux-gnu/libcudnn_cnn_train.so.8.1.1
/usr/lib/x86_64-linux-gnu/libcudnn_ops_infer.so.8.1.1
/usr/lib/x86_64-linux-gnu/libcudnn_ops_train.so.8.1.1
HIP runtime version: N/A
MIOpen runtime version: N/A

Versions of relevant libraries:
[pip3] numpy==1.18.5
[pip3] torch==1.7.1+cu110
[pip3] torchaudio==0.7.2
[pip3] torchgeometry==0.1.2
[pip3] torchvision==0.8.2+cu110
[conda] Could not collect

## Additional context

cc @ezyang @gchanan @zou3519 @mruberry @rgommers @heitorschueroff

### PR DESCRIPTION
**Pull request by peterbell10**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #79393

Closes #54789

For the `fused_mode` kernel, this just uses `c10::load` but the
`apply_mode` function is a bit harder because it uses `thrust`.
Instead, I've added a second dedicated path for bool which also
only uses 2 thrust calls instead of the normal 6, by exploiting
the fact that bools only have two possible values.

In the following `timeit` benchmark which calls the `apply_mode`
version, I see execution time drop from 16.9 ms to 2.2 ms (which is
still terrible, but my main goal is fixing the bool handling).

```python
import torch a = torch.randint(
    0, 2, size=(100, 4096), device='cuda', dtype=torch.bool)
%timeit a.mode(1)
```

### CODE REVIEW

File: aten/src/ATen/native/cuda/TensorModeKernel.cu [hunk 1]
```diff
@@ -19,16 +21,113 @@
...
+    at::cuda::ThrustAllocator thrust_allocator;
+    auto stream = at::cuda::getCurrentCUDAStream();
+    auto policy = thrust::cuda::par(thrust_allocator).on(stream);
+
+    // For bool, we can skip finding the unique elements since there
+    // are only two possible values.
+
+    // See NOTE [Loading boolean values]
+    auto first_bytes = reinterpret_cast<const uint8_t*>(first);
+    auto last_bytes = reinterpret_cast<const uint8_t*>(first);
```
Comment (ngimel) [inline aten/src/ATen/native/cuda/TensorModeKernel.cu:100]: should this be `last`? How come tests are passing?

Comment (peterbell10) [reply aten/src/ATen/native/cuda/TensorModeKernel.cu:100]: After some investigation, this `apply_mode` path is only tested for int64 and is completely broken for all other dtypes even without this PR.

Comment (ngimel) [reply aten/src/ATen/native/cuda/TensorModeKernel.cu:100]: Should we throw an error for other dtypes then? This path is very rarely used, so it shouldn't be a big deal.

Comment (peterbell10) [reply aten/src/ATen/native/cuda/TensorModeKernel.cu]: It's just a small typo, but it critically brakes the code. I've opened #81898 to fix it and add more tests.

**Review by ngimel — APPROVED** [review]

**Comment by janeyx99** [comment]
@pytorchbot revert -m "Sorry, reverting as it breaks 10.2 build run only on trunk https://hud.pytorch.org/pytorch/pytorch/commit/7592d3f6464389a2fef02474710a99f3729ca057" -c nosignal

Sorry about that! Please include ciflow/trunk label when relanding this PR. We also have a new `@pytorchbot merge -l` which will run all ciflow/trunk jobs during the land process that you can try out! https://github.com/pytorch/pytorch/wiki/Bot-commands
